server/src/main/java/org/elasticsearch/search/aggregations/metrics/GlobalOrdCardinalityAggregator.java

```diff
@@ -0,0 +1,203 @@
+    @Nullable
+    private HyperLogLogPlusPlusSparse counts;
+
+    private OrdinalsCollector collector;
+
+    private final int precision;
+
+    public GlobalOrdCardinalityAggregator(
+            String name,
+            ValuesSource.Bytes.WithOrdinals valuesSource,
+            int precision,
+            int maxOrd,
+            SearchContext context,
+            Aggregator parent,
+            Map<String, Object> metadata) throws IOException {
+        super(name, context, parent, metadata);
+        this.valuesSource = valuesSource;
+        this.counts = null;
+        this.collector = new OrdinalsCollector(context.bigArrays(), maxOrd);
+        this.precision = precision;
+    }
+
+    @Override
+    public ScoreMode scoreMode() {
+        return valuesSource.needsScores() ? ScoreMode.COMPLETE : ScoreMode.COMPLETE_NO_SCORES;
+    }
+
+    @Override
+    public LeafBucketCollector getLeafCollector(LeafReaderContext ctx,
+            final LeafBucketCollector sub) throws IOException {
+        this.collector.set(valuesSource.globalOrdinalsValues(ctx));
+        return collector;
+    }
+
+    @Override
+    protected void doPostCollection() throws IOException {
+        counts = new HyperLogLogPlusPlusSparse(precision, collector.bigArrays, collector.visitedOrds.size());
+        collector.postCollect(counts);
+        collector.close();
+        collector = null;
+    }
+
+    @Override
+    public double metric(long owningBucketOrd) {
+        return counts == null ? 0 : counts.cardinality(owningBucketOrd);
+    }
+
+    @Override
+    public InternalAggregation buildAggregation(long owningBucketOrdinal) {
+        if (counts == null || owningBucketOrdinal >= counts.maxOrd() || counts.cardinality(owningBucketOrdinal) == 0) {
+            return buildEmptyAggregation();
+        }
+        // We need to build a copy because the returned Aggregation needs remain usable after
+        // this Aggregator (and its HLL++ counters) is released.
+        AbstractHyperLogLogPlusPlus copy = counts.clone(owningBucketOrdinal, BigArrays.NON_RECYCLING_INSTANCE);
+        return new InternalCardinality(name, copy, metadata());
+    }
+
+    @Override
+    public InternalAggregation buildEmptyAggregation() {
+        return new InternalCardinality(name, null, metadata());
+    }
+
+    @Override
+    protected void doClose() {
+        Releasables.close(counts, collector);
+    }
+
+    @Override
+    public void collectDebugInfo(BiConsumer<String, Object> add) {
+        // maybe need to add something for global?
+        super.collectDebugInfo(add);
+        add.accept("empty_collectors_used", 0);
+        add.accept("numeric_collectors_used", 0);
+        add.accept("ordinals_collectors_used", 1);
+        add.accept("ordinals_collectors_overhead_too_high", 0);
+        add.accept("string_hashing_collectors_used", 0);
+    }
+
+    private static class OrdinalsCollector extends LeafBucketCollector implements Releasable {
```

Comment:
I wonder if we need a second class here - maybe all of these members should just go into the top level class?

Reply:
+1

Reply:
I have tried and I am not convince we should create that layer.

Reply:
I was suggesting we remove this layer. Is that what you tried and found wasn't worth it?

Reply:
I tried to create a base class for the cardinality aggregators and move collectors to that layer, but because the flow on the ordinals collector is different to the standard collector, it gives more issues than solve them.

Reply:
Oh! I was just thinking that the members on this class could just be moved to the outer class and that we could return a super simple collector like we do for most aggs. Is that worth it, do you think?

Reply:
Ooops, I understand, I did the change, it makes sense.